dot/rpc/modules/state.go

```diff
@@ -403,30 +404,71 @@ func (sm *StateModule) GetStorageSize(
 	return nil
 }
 
-// QueryStorage isn't implemented properly yet.
+// QueryStorage Query historical storage entries (by key) starting from a request start block given,
+//until request end block or best block if end block in nil
 func (sm *StateModule) QueryStorage(
 	_ *http.Request, req *StateStorageQueryRangeRequest, res *[]StorageChangeSetResponse) error {
 	if req.StartBlock.IsEmpty() {
-		return errors.New("the start block hash cannot be an empty value")
+		return ErrStartBlockHashEmpty
 	}
 
-	changesByBlock, err := sm.coreAPI.QueryStorage(req.StartBlock, req.EndBlock, req.Keys...)
+	startBlock, err := sm.blockAPI.GetBlockByHash(req.StartBlock)
 	if err != nil {
 		return err
 	}
 
-	response := make([]StorageChangeSetResponse, 0, len(changesByBlock))
+	startBlockNumber := startBlock.Header.Number
 
-	for block, c := range changesByBlock {
-		var changes [][]string
+	var endBlockHash common.Hash
+	if req.EndBlock.IsEmpty() {
+		endBlockHash = sm.blockAPI.BestBlockHash()
+	} else {
+		endBlockHash = req.EndBlock
+	}
+	endBlock, err := sm.blockAPI.GetBlockByHash(endBlockHash)
+	if err != nil {
+		return fmt.Errorf("cannot get block by hash: %w", err)
+	}
+	endBlockNumber := endBlock.Header.Number
 
-		for key, value := range c {
-			changes = append(changes, []string{key, value})
+	response := make([]StorageChangeSetResponse, 0)
+	lastValue := make([]*string, len(req.Keys))
+	firstPass := true
+	for i := startBlockNumber; i <= endBlockNumber; i++ {
+		blockHash, err := sm.blockAPI.GetHashByNumber(i)
+		if err != nil {
+			return fmt.Errorf("cannot get hash by number: %w", err)
 		}
+		var changes [][]*string
 
-		blochHash := block
+		for j, key := range req.Keys {
+			value, err := sm.storageAPI.GetStorageByBlockHash(&blockHash, common.MustHexToBytes(key))
+			if err != nil {
+				return fmt.Errorf("cannot get value by block hash: %w", err)
+			}
+			var hexValue *string
+			if len(value) > 0 {
+				h := common.BytesToHex(value)
+				hexValue = &h
+			}
+			k := key
+			if firstPass {
+				changes = append(changes, []*string{&k, hexValue})
+				lastValue[j] = hexValue
+			} else {
+				if lastValue[j] == nil && hexValue != nil {
+					changes = append(changes, []*string{&k, hexValue})
+					lastValue[j] = hexValue
+				}
+				if lastValue[j] != nil && *lastValue[j] != *hexValue {
+					changes = append(changes, []*string{&k, hexValue})
+					lastValue[j] = hexValue
+				}
+			}
```

Comment:
nit use `continue` to reduce indentation of the second block

```suggestion
				continue
			}
			if lastValue[j] == nil && hexValue != nil {
				changes = append(changes, []*string{&k, hexValue})
				lastValue[j] = hexValue
			}
			if lastValue[j] != nil && *lastValue[j] != *hexValue {
				changes = append(changes, []*string{&k, hexValue})
				lastValue[j] = hexValue
			}
```